blockcache/blockcache.go

```diff
@@ -0,0 +1,224 @@
+			return nil, err
+		}
+	}
+
+	path := filepath.Join(dbPath, dbName)
+	bdb, err := bolt.Open(path, dbFilePermission, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the bucket if necessary.
+	err = bdb.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucket(blockCacheBucket)
+
+		if err != nil && err != bolt.ErrBucketExists {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to create cache db")
+	}
+
+	return &MostRecentBlockCache{
+		db:           bdb,
+		capacity:     capacity,
+		blockHeaders: blockHeaders,
+	}, nil
+}
+
+// Close instructs the underlying database to stop cleanly.
+func (c *MostRecentBlockCache) Close() {
+	c.db.Close()
+}
+
+// PutBlock attempts to insert the block into the cache. If the cache is at
+// capacity it will first attempt to evict the oldest block, if the given block
+// is older than all blocks currently in the cache, then no action happens.
+func (c *MostRecentBlockCache) PutBlock(block *wire.MsgBlock) error {
+	// The key for this block is its height in the blockchain.
+	// This allows us to evict the oldest entry easily.
+	hash := block.BlockHash()
+	height, err := c.blockHeaders.HeightFromHash(&hash)
+	if err != nil {
+		return err
+	}
+
+	key := []byte(strconv.FormatUint(uint64(height), 10))
+
+	// Serialize the block.
+	w := bytes.NewBuffer(make([]byte, 0, block.SerializeSize()))
+	err = block.Serialize(w)
+	if err != nil {
+		return err
+	}
+
+	// Insert the block, evicting the oldest block if the capacity has been
+	// reached.
+	err = c.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(blockCacheBucket)
+
+		isFull := bucket.Stats().KeyN >= c.capacity
+
+		if isFull {
+			// Find the oldest block we have in the cache.
+			oldestKey, _ := bucket.Cursor().First()
+			isNewer := bytes.Compare(key, oldestKey) == 1
+
+			// Only insert this block if it wouldn't evict a block that's newer
+			// than itself.
+			if isNewer {
+				bucket.Delete(oldestKey)
+				bucket.Put(key, w.Bytes())
+			}
+		} else {
+			bucket.Put(key, w.Bytes())
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PrintStats prints the cache hit rate to standard output.
+func (c *MostRecentBlockCache) PrintStats() {
+	hits := float64(atomic.LoadUint64(&c.hits))
+	misses := float64(atomic.LoadUint64(&c.misses))
+	fmt.Printf("hits: %.0f, total: %.0f (%.01f)\n", hits, hits+misses, hits/(hits+misses))
+}
+
+// FetchBlock queries the cache for a block matching the given hash, returning
+// the block or ErrBlockNotFound if it couldn't be retrieved.
+func (c *MostRecentBlockCache) FetchBlock(hash *chainhash.Hash) (*wire.MsgBlock,
+	error) {
+	// Look up the height from the hash, then query by height and double
+	// check we got the right block.
+	height, err := c.blockHeaders.HeightFromHash(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	key := []byte(strconv.FormatUint(uint64(height), 10))
```

Comment:
Similar comment here about using a compact encoding.

Reply:
Actually, when moving to a size based cache, this goes away all together as the key will likely be the hash of the block.